src/Altinn.App.Core/Features/Action/SigningUserAction.cs

```diff
@@ -92,17 +100,39 @@ public async Task<UserActionResult> HandleAction(UserActionContext context)
         );
     }
 
+    private static bool ShouldSign(
+        ProcessTask currentTask,
+        List<DataElement> dataElements,
+        List<DataType>? dataTypesToSign
+    )
+    {
+        var signatureIsConfigured =
+            currentTask.ExtensionElements?.TaskExtension?.SignatureConfiguration?.SignatureDataType is not null;
+        if (dataTypesToSign.IsNullOrEmpty() || !signatureIsConfigured)
+        {
+            return false;
+        }
+
+        var dataElementMatchExists = dataElements.Any(de =>
+            dataTypesToSign!.Any(dt => string.Equals(dt.Id, de.DataType, StringComparison.OrdinalIgnoreCase))
```

Comment:
I think we should avoid null-forgiving operator as much as possible. In this case, using `NotNullWhenAttribute` on `IsNullOrEmpty` might be enough for it to realize it's not null?

Where we can't help the compiler with the flow analysis, I think it's better to either use
`Debug.Assert` or a manually thrown exception with a message describing the assumption rather than ending up with nullref

Reply:
Whether or not we should use null-forgiving operators in cases where the compilers flow analysis "fails" should perhaps be a team rule?
I lean towards not using it, simply because it can be hard to spot

Reply:
yeah agreed!